### PR TITLE
feat(website): store options TypeScript, Enable jsx and AST Viewer in browser's local storage

### LINKED ISSUE #5754: Website: Remember AST Viewer selection if URL is wiped
### Before You File a Documentation Request Please Confirm You Have Done The Following...

- [X] I have looked for existing [open or closed documentation requests](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Adocumentation) that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Suggested Changes

I've been using https://typescript-eslint.io/play as a replacement for astviewer.net because it also shows lint & TypeScript messages. But, I keep having to manually change the _AST Viewer_ option from the default _Disabled_ value to _ESTree_. That's irksome.

It'd be nice if changing AST Viewer also saved to `localStorage`, so that when https://typescript-eslint.io/play is next visited, the previous selection is remembered.

### Affected URL(s)

https://typescript-eslint.io/play

### PR DESCRIPTION
**Pull request by DKurilo**
## PR Checklist

- [x] Addresses an existing open issue: fixes #5754 
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

Each time when website configuration is changed it updates hash in url. After this PR it will also try to store data in local storage and when web site is opened it will take default parameters, apply parameters from local storage (if they exists) and then apply parameters from hash. This way TypeScript version, showAST, and enable jsx will be stored locally:.

### CODE REVIEW

File: packages/website/src/components/hooks/useHashState.ts [hunk 1]
```diff
@@ -114,16 +115,69 @@ const writeStateToUrl = (newState: ConfigModel): string => {
...
+    }
+
+    const config: unknown = JSON.parse(configString);
+    if (typeof config !== 'object' || !config) {
+      return undefined;
+    }
+
+    const state: Partial<ConfigModel> = {};
+    if (hasOwnProperty('ts', config)) {
+      const ts: unknown = config.ts;
```
Comment (JoshuaKGoldberg) [inline packages/website/src/components/hooks/useHashState.ts:132]: Nit: no need for this type annotation:

```suggestion
      const ts = config.ts;
```

File: packages/website/src/components/lib/has-own-property.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+export function hasOwnProperty<X extends object, Y extends PropertyKey>(
```
Comment (JoshuaKGoldberg) [inline packages/website/src/components/lib/has-own-property.ts:1]: Lol. What a great example of why TypeScript's `in` type narrowing sometimes falls short.

(not requesting changes, just griping about the world we live in)

**Review by JoshuaKGoldberg — DISMISSED** [review]
Fantastic, thanks @DKurilo! This works great, nicely done. 🔥 

Just a few little nitpicks from me that I can apply before merging.

File: packages/website/src/components/hooks/useHashState.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (DKurilo) [reply packages/website/src/components/hooks/useHashState.ts]: Thanks.
Yeah, I used to write type annotations everywhere (even in Haskell that has a lot better inference system) and I just forgot to remove it. :)

File: packages/website/src/components/lib/has-own-property.ts (hunk 2 above)
Comment (DKurilo) [reply packages/website/src/components/lib/has-own-property.ts]: Yeah, as I remember, current TS version can infer type from using `in`.

**Review by JoshuaKGoldberg — APPROVED** [review]

**Pull request merged**